wandering-warriors/modules/ledger.py

```diff
@@ -0,0 +1,190 @@
+from kivy.uix.boxlayout import BoxLayout
+from kivy.uix.label import Label
+from kivy.uix.popup import Popup
+
+import pandas as pd
+
+
+class Ledger(BoxLayout):
+    """
+    Ledger data structure:
+    x: first number in equation (float)
+    y: second number in equation (float)
+    op: operator ['+', "-", '*', '/'] (str)
+    z: result (float)
+    """
+
+    def __init__(self, **kwargs):
+        super(Ledger, self).__init__(**kwargs)
+        self.col = 'x'
+        self.row = 1
+        self.df = pd.DataFrame(columns=['x', 'y', 'op', 'z'])
+        self.df.loc[1] = {'x': 0, 'y': 0, 'op': '', 'z': 0}
+        self.clear_button_src = 'assets/graphics/clear.png'
+
+    def select(self, col: str):
+        """select active column"""
+        if col not in ['x', 'y', 'op', 'z']:
+            print(f"[ WARNING ] Invalid column: {col}")
+            pass
+        else:
+            self.col = col
+
+    def update(self, n: int, op: str = '='):
+        """update active cell"""
+        if op == '+':
+            self.df.at[self.row, self.col] += n
+        if op == '-':
+            self.df.at[self.row, self.col] -= n
+        if op == '*':
+            self.df.at[self.row, self.col] *= n
+        if op == '/':
+            self.df.at[self.row, self.col] /= n
+        if op == '=':
+            self.df.at[self.row, self.col] = n
+        self.refresh_ledger()
+
+    def open_help(self):
+        TEXT = '''
+            Input digits in the sandbox to tabulate numbers on the ledger.
+            You can perform operations on numbers using the operations buttons immediately to the right.
+        '''
+
+        p = Popup(
+            title='Ledger Help',
+            size_hint_y=None,
+            height=200,
+            content=Label(text=TEXT)
+        )
+
+        p.open()
+
+    def operation(self, op: str):
+        """update operator column of current row and advance selection"""
+        if op not in ['+', '-', '*', '/']:
+            print(f"[ WARNING ] Invalid operator: {op}")
```

Comment:
This print seems to imitate what a logger powered by Python's `logging` module would output. I'd recommend you to look into the `logging` module and start utilizing it for these type of tasks. This will allow you to specify logging levels (which means you don't even have to fully remove debug information since those debug message actually be useful during debugging!) and save your logs to file instead of STDERR/STDOUT if you so wish. Once you get used to it, it's a very flexible system that integrates well with broader logging frameworks.